mllib/src/test/scala/org/apache/spark/ml/regression/RandomForestRegressorSuite.scala

```diff
@@ -105,6 +108,55 @@ class RandomForestRegressorSuite extends SparkFunSuite with MLlibTestSparkContex
       }
   }
 
+  test("Random Forest variance") {
+    val categoricalFeatures = Map.empty[Int, Int]
+    val df: DataFrame = TreeTests.setMetadata(
+      orderedLabeledPoints50_1000, categoricalFeatures, 0)
+
+    // RF with one tree should have the same variance as that of the tree.
+    val rf = new RandomForestRegressor()
+      .setImpurity("variance")
+      .setMaxDepth(30)
+      .setNumTrees(1)
+      .setMaxBins(10)
+      .setFeatureSubsetStrategy("all")
+      .setSubsamplingRate(1.0)
+      .setSeed(123)
+
+    val rfModel = rf.fit(df)
+    val rfVariances = rfModel.transform(df).select("variance").collect()
+
+    val dt = new DecisionTreeRegressor()
+      .setImpurity("variance")
+      .setMaxDepth(30)
+      .setMaxBins(10)
+      .setSeed(123)
+    val dtModel = dt.fit(df)
+    val dtVariances = dtModel.transform(df).select("variance").collect()
+    val nSamples = dtVariances.size
+    (0 to nSamples - 1).foreach { i =>
+      val diff = math.abs(rfVariances(i).getDouble(0) - dtVariances(i).getDouble(0))
+      assert(diff < 1e-6)
+    }
+
+    rf.setMaxDepth(2)
+    rf.setNumTrees(20)
+    val rfNewModel = rf.fit(df)
+    val results = rfNewModel.transform(df).select("features", "variance").collect()
+    val features = col("features")
+    val trees = rfNewModel.trees
+    val numTrees = rfNewModel.getNumTrees
+    results.map { case Row(features: Vector, variance: Double) =>
+      val rootNodes = trees.map(_.rootNode.predictImpl(features))
+      val predsquared = rootNodes.map(x => math.pow(x.prediction, 2)).sum / numTrees
+      val variance = rootNodes.map(_.impurityStats.calculate()).sum / numTrees
+      val predictions = rootNodes.map(_.prediction).sum / numTrees
+      val expectedVariance = -math.pow(predictions, 2) + variance + predsquared
+      assert(variance === expectedVariance,
+        s"Expected variance $expectedVariance but got $variance.")
```

Comment:
I believe you are using the wrong variance here, since you re-use that variable name.